### ISSUE TITLE
Support for dynamic component types

### ISSUE THREAD

**Issue by mrZalli** [slide-rs/specs#64]
This library seems promising and I would like to use it in a project of mine, but I really can't because I would like the ability to create custom component types at runtime.

The current implementation seems like it uses typeids to map component types into their relative data. It seems like it would be easy to change them from typeids into data of generic type that implements the necessary traits (like `PartialEq`, `Eq`, `Clone`, etc).

The problem is that I don't really know how to make one implementation that supports the current way of accessing components and this dynamic one at the same time. There could be two different `World` implementations with slightly different interfaces but that would be messy and unwanted.

I think I'll fork this project to try and make a dynamic version and see what I can do.

**Comment by kvark** [comment]
Hi @mrZalli and thanks for filing this issue!
Would you mind sharing the details of your use-case, e.g. why and where do you need runtime-defined components?

**Comment by mrZalli** [comment]
Thanks for the fast response!

The project I'm doing (or planning to do) is a sort of a game engine, where most of the content and some rules of the world are in external text files to allow easy editing and modding.

Of course I could allow only a pre-existing selection of components and it would work, but I feel like it would be an arbitrary restriction. Because of that I would like if new components could be introduced in the aforementioned text files when the program starts.

**Comment by kvark** [comment]
> where most of the content and some rules of the world are in external text files to allow easy editing and modding

But how do you process these run-time entities then? There has to be some code doing stuff with them, and dealing with an untyped memory blob doesn't sound too Rusty for the lack of type safety.

**Comment by mrZalli** [comment]
These components values can be enums with different variants for different datatypes or something like JSON strings.

The rules that refer to these components are also defined in text files and could be written in some script language, like Lua.

**Comment by mrZalli** [comment]
Now when I think of this I have a solution of sorts. You can just add another (optional) identifying value alongside the the type-id and it should work. Then you could have additional functions that take these dynamic identifiers as their parameters.
I'll try to do something like this in a fork.

**Comment by OvermindDL1** [comment]
This is the same thing I spoke of before.  Remember when I mentioned that in my old engine that you can define new component types in LUA (type names are a 64-bit integer) and they could define systems over them (or even from C++, or vice-versa, as Aspects work based on the 64-bit integer type IDs).  It is a fairly important feature for moddability.

**Comment by mrZalli** [comment]
So I made a simple support for the dynamic components, but it's still kinda clunky and not tested yet.
https://github.com/mrZalli/specs

**Comment by mrZalli** [comment]
I think I have to add that I edited ’World’ and ’EntityBuilder’ so that their old interface still works even though I ran into some problems with type inference when using the ’new’ function.

**Comment by csherratt** [comment]
@mrZalli looking over your changes it looks to be a net-win without costing us to much complexity. Since we can `opt-in` to using the feature. There are some minor issues that are just a result of rust not being annoying with how borrows work.

For scripted types this makes quite a bit of sense. I could see a few cases where you might want to use both in your engine. It could be worked around by abusing the `Default` type.

**Comment by mrZalli** [comment]
@csherratt that's great. The way I created this is probably not the best one, because it allows for components with the same component id names, but different datatypes, which is probably confusing and unwanted.

A more simpler and cleaner design would allow only one type of identification method used at the same time. Could this be done with something like two-variant enum keys for the `World`'s `components` hash map? Or just opt-inning features.

What the enum key type might look like:
```Rust
enum CompKey<T: PartialEq + Eq + Hash> {
    Type(TypeId),
    CompId {
        id: T,
        type: TypeId
    }
}
```
(The implementation of PartialEq and Hash for that enum type has to be done manually.)

PS: I fixed some minor flaws and errors and added a simple test case to my fork.
